test/assign_dna_unit_test.dart

```diff
@@ -0,0 +1,1202 @@
+      */
+
+      var helices = [Helix(idx: 0, max_offset: 10, grid: Grid.square)];
+      var design = Design(helices: helices, grid: Grid.square);
+
+      design = design.strand(0, 0).move(5).commit();
+      design = design.strand(0, 5).move(-5).commit();
+
+      AppState state = app_state_from_design(design);
+      Strand strand = design.strands.last;
+      String dna_sequence = 'AAAAC';
+
+      state = app_state_reducer(
+          state,
+          AssignDNA(
+            strand: strand,
+            assign_complements: true,
+            dna_sequence: dna_sequence,
+            warn_on_change: true,
+          ));
+      expect(state.design.strands[0].dna_sequence, 'GTTTT');
+    });
+
+    test('AssignDNA__assign_seq_with_wildcards', () {
+      /*
+        0   4
+        G??TT
+        [--->
+        C??AA
+        <---]
+      */
+
+      var helices = [Helix(idx: 0, max_offset: 10, grid: Grid.square)];
+      var design = Design(helices: helices, grid: Grid.square);
+
+      design = design.strand(0, 0).move(5).commit();
+      design = design.strand(0, 5).move(-5).commit();
+
+      AppState state = app_state_from_design(design);
+      Strand strand = design.strands.last;
+      String dna_sequence = 'AA??C';
+
+      state = app_state_reducer(
+          state,
+          AssignDNA(
+            strand: strand,
+            assign_complements: true,
+            dna_sequence: dna_sequence,
+            warn_on_change: true,
+          ));
+      expect(state.design.strands[0].dna_sequence, 'G??TT');
+    });
+
+    test('AssignDNA__one_strand_assigned_by_complement_from_two_other_strands', () {
+      /*
+        0      8
+        TTTGCCCT
+        [------>
+        CAAAAGGG
+        <--]<--]
+      */
+
+      var helices = [Helix(idx: 0, max_offset: 10, grid: Grid.square)];
+      var design = Design(helices: helices, grid: Grid.square);
+
+      design = design.strand(0, 0).move(8).commit();
+      design = design.strand(0, 4).move(-4).commit();
+      design = design.strand(0, 8).move(-4).commit();
+
+      AppState state = app_state_from_design(design);
+      Strand strand1 = design.strands[1];
+      Strand strand2 = design.strands.last;
+      String dna_sequence1 = 'CAAA';
+      String dna_sequence2 = 'AGGG';
+
+      state = app_state_reducer(
+          state,
+          AssignDNA(
+            strand: strand1,
+            assign_complements: true,
+            dna_sequence: dna_sequence1,
+            warn_on_change: true,
+          ));
+      expect(state.design.strands[0].dna_sequence, 'TTTG????');
+      state = app_state_reducer(
+          state,
+          AssignDNA(
+            strand: strand2,
+            assign_complements: true,
+            dna_sequence: dna_sequence2,
+            warn_on_change: true,
+          ));
+      expect(state.design.strands[0].dna_sequence, 'TTTGCCCT');
+    });
+
+    test('AssignDNA__adapter_assigned_from_scaffold_and_tiles', () {
+      /*
+          0    5    10    
```

Comment:
Some of this ASCII art is a little tough to read. For instance, this one looks like the crossover from the top-most strand might be going down to the bottom-most strand, through the strand in the middle, but I'm not sure.

I know I said the convention doesn't matter as long as it's readable, but this isn't very readable to me. Maybe it would be better to use the convention from the Python comments in this case, where bases ACGT represent where DNA is, and dashes represent empty space, so that the crossover from the topmost strand to the bottom most strand is "visible" through the ASCII art. and doesn't "disappear behind" anything. (Or another convention that doesn't hide any crossovers.)

I didn't comment explicitly on any other ASCII art, but same applies to all of them.